### ISSUE TITLE
Test failures when compiled with GCC graphite flags

### ISSUE THREAD

**Issue by orlitzky** [flintlib/flint#2029]
Original report: https://bugs.gentoo.org/934463

I'm able to confirm this. When combined with `-O2` or `-O3`, some of GCC's graphite optimizations leads to crashes. On my machine the baseline `$CFLAGS` are `-O2 -pipe -march=native`, and flint's test suite passes with those. If I append `-floop-nest-optimize` (disclaimer: this flag is experimental) to those, then I get a failure:

```
        Start   7: src-d_vec-test-main
  4/123 Test   #7: src-d_vec-test-main ....................***Exception: SegFault  0.04 sec
```

If instead I append `-fgraphite-identity` to the baseline, I get several more failures:

```
        Start  22: src-fmpz_mod_poly-test-main
 19/123 Test  #22: src-fmpz_mod_poly-test-main ............Subprocess aborted***Exception:   0.30 sec

        Start  19: src-fmpz_poly-test-main
 23/123 Test  #19: src-fmpz_poly-test-main ................Subprocess aborted***Exception:  14.15 sec

        Start  66: src-fmpz_mpoly_factor-test-main
 63/123 Test  #66: src-fmpz_mpoly_factor-test-main ........Subprocess aborted***Exception:   0.07 sec

        Start 108: src-gr-test-main
 121/123 Test #108: src-gr-test-main .......................Subprocess aborted***Exception: 383.36 sec

        Start 117: src-qqbar-test-main
 122/123 Test #117: src-qqbar-test-main ....................Subprocess aborted***Exception: 835.32 sec
```

I would expect `-fgraphite-identity` to be "safe," but I'm no expert.

**Comment by fredrik-johansson** [comment]
I'm unable to reproduce the ``-floop-nest-optimize`` issue with GCC 12.3.

Compiling with ``-O2 -pipe -march=native -fgraphite-identity``:

* The ``fmpz_mod_poly`` tests pass

* ``fmpz_poly`` tests pass, but fail with ``FLINT_TEST_MULTIPLIER=10``

* ``fmpz_mpoly_factor`` tests fail

* ``gr`` and ``qqbar`` test seem to hang

**Comment by fredrik-johansson** [comment]
This might be the easiest failure to debug:

```
$ build/fmpz_poly/test/main fmpz_poly_taylor_shift

FAIL
3  188009194635 44888971 1
2  1 1
3  188009194635 44888971 1
3  188054083607 44888973 1
```

Here we want to Taylor shift the length-3 polynomial ``[188009194635, 44888971, 1]`` by 1. The expected result is ``[188054083607, 44888973, 1]`` but we get back the original input unchanged.

This goes through the basecase code in ``fmpz_poly/taylor_shift_horner.c`` where we run the following loop:

```
        for (i = n - 2; i >= 0; i--)
            for (j = i; j < n - 1; j++)
                poly[j] += poly[j + 1];
```

So unless I'm missing something obvious here, GCC miscompiles this loop when graphite is enabled.

**Comment by fredrik-johansson** [comment]
Reduced test case for GCC:

```
#include <stdio.h>

void
shift(long * c, long n)
{
    long i, j;

    if (n <= 62)
    {
        for (i = n - 2; i >= 0; i--)
            for (j = i; j < n - 1; j++)
                c[j] += c[j + 1];
    }
}

int main(int argc, char ** argv)
{
    long c[3];
    volatile long n = 3;
    c[0] = 3;
    c[1] = 2;
    c[2] = 1;
    shift(c, n);
    printf("%ld %ld %ld\n", c[0], c[1], c[2]);
}
```

```
$ gcc-12 -O2 graphite.c
$ ./a.out
6 4 1

$ gcc-12 -O2 -fgraphite-identity graphite.c
$ ./a.out
3 2 1
```

**Comment by orlitzky** [comment]
> Definitely looks like a GCC bug. Isolated test case:
> 
> ```
> $ gcc-12 -O2 graphite.c
> $ ./a.out
> 6 4 1
> 
> $ gcc-12 -O2 -fgraphite-identity graphite.c
> $ ./a.out
> 3 2 1
> ```

Interesting, I get the same with gcc-13.2. I'll try to build gcc from git and see if it persists, thanks.

**Comment by orlitzky** [comment]
Same thing with the latest GCC: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=115697

Closing as not a flint problem. Thanks for the test case!
